File: benchmark/benchmark.py
import math
import time
from typing import Callable, List, Union

ITERATIONS = 10000


def benchmark(callback: Callable[[], Union[str, int]]):
    run_times: List[str] = []

    answer: Union[str, int] = 0

    for x in range(ITERATIONS):
        start_at = time.time()
        answer = callback()
        end_at = time.time()
        run_times.append((end_at - start_at) * 1000)

    run_times.sort()

    total = sum(run_times)
    average = total / ITERATIONS
    median = run_times[math.floor(ITERATIONS / 2)]
    p95 = run_times[math.floor(ITERATIONS * 0.95)]
    p99 = run_times[math.floor(ITERATIONS * 0.99)]

    print(
        {
            "answer": answer,
            "total": parse_result(total),
            "average": parse_result(average),
            "median": parse_result(median),
            "p95": parse_result(p95),
            "p99": parse_result(p99),
        }
    )


def parse_result(result: int):
    return "{:.5f}ms".format(result)

File: benchmark/benchmark.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.benchmark = void 0;
const fs_1 = __importDefault(require("fs"));
const ITERATIONS = 10_000;
function benchmark(callback, key) {
    const runTimes = [];
    let answer = 0;
    for (let i = 0; i < ITERATIONS; i++) {
        const startAt = performance.now();
        answer = callback();
        const endAt = performance.now();
        runTimes.push(endAt - startAt);
    }
    runTimes.sort();
    const total = runTimes.reduce((sum, time) => sum + time, 0);
    const average = total / ITERATIONS;
    const median = runTimes[Math.floor(ITERATIONS / 2)];
    const p95 = runTimes[Math.floor(ITERATIONS * 0.95)];
    const p99 = runTimes[Math.floor(ITERATIONS * 0.99)];
    const results = {
        answer,
        total: parseResult(total),
        average: parseResult(average),
        median: parseResult(median),
        p95: parseResult(p95),
        p99: parseResult(p99),
    };
    console.log(results);
    fs_1.default.writeFileSync(`./results/${key}_ts.json`, JSON.stringify(results));
}
exports.benchmark = benchmark;
function parseResult(result) {
    return `${result.toFixed(5)}ms`;
}
